# This time: with all the data.

In [2]:
path = "Resources/Valve_Player_Data.csv"

import matplotlib.pyplot as plt
import pandas as pd
import requests
import numpy as np
from pprint import pprint
import time
import datetime
from dateutil.relativedelta import relativedelta

In [3]:
monthly_players = pd.read_csv(path)
monthly_players

Month_Year  Avg_players       Gain Percent_Gain  Peak_Players  \
0     September 2021    512350.92     268.96       +0.05%        942519   
1        August 2021    512081.96    6014.60       +1.19%        802544   
2          July 2021    506067.36  -43279.72       -7.88%        763523   
3          June 2021    549347.08 -110541.81      -16.75%        929940   
4           May 2021    659888.89  -63457.63       -8.77%       1087197   
...              ...          ...        ...          ...           ...   
5266   December 2016       192.03     -21.90      -10.24%           405   
5267   November 2016       213.92    -134.68      -38.63%           537   
5268    October 2016       348.60    -201.75      -36.66%          1031   
5269  September 2016       550.36     543.26    +7654.89%          2836   
5270     August 2016         7.10        NaN          NaN            18   

                                     URL        Date  \
0        https://steamcharts.com/app/730  2021-09-01   
1        https://steamcharts.com/app/730  2021-08-01   
2        https://steamcharts.com/app/730  2021-07-01   
3        https://steamcharts.com/app/730  2021-06-01   
4        https://steamcharts.com/app/730  2021-05-01   
...                                  ...         ...   
5266  https://steamcharts.com/app/435150  2016-12-01   
5267  https://steamcharts.com/app/435150  2016-11-01   
5268  https://steamcharts.com/app/435150  2016-10-01   
5269  https://steamcharts.com/app/435150  2016-09-01   
5270  https://steamcharts.com/app/435150  2016-08-01   

                             Game_Name  
0     Counter Strike: Global Offensive  
1     Counter Strike: Global Offensive  
2     Counter Strike: Global Offensive  
3     Counter Strike: Global Offensive  
4     Counter Strike: Global Offensive  
...                                ...  
5266          Divinity: Original Sin 2  
5267          Divinity: Original Sin 2  
5268          Divinity: Original Sin 2  
5269          Divinity: Original Sin 2  
5270          Divinity: Original Sin 2  

[5271 rows x 8 columns]

In [4]:
monthly_players.nunique()

Month_Year       111
Avg_players     5241
Gain            5135
Percent_Gain    3685
Peak_Players    4988
URL               98
Date             111
Game_Name         98
dtype: int64

In [5]:
monthly_players.dtypes

Month_Year       object
Avg_players     float64
Gain            float64
Percent_Gain     object
Peak_Players      int64
URL              object
Date             object
Game_Name        object
dtype: object

In [6]:
# EXTRACT AND TRANSFORM DATA

# drop first two months for every game as these extreme values will totally skew the data.

In [7]:
monthly_players.Game_Name.unique()

array(['Counter Strike: Global Offensive', 'Dota 2',
       'PUBG: Battlegrounds', 'Apex Legends', 'Team Fortress 2',
       'Grand Theft Auto V', 'NARAKA: BLADEPOINT', 'Wallpaper Engine',
       'Rust', 'Destiny 2', 'Warframe', 'Cookie Clicker',
       "Tom Clancy's Rainbow Six Seige", 'Football Manager 2021 ',
       'PAYDAY 2', 'ARK: Survival Evolved', 'Dead by Daylight',
       "Sid Meier's Civilization VI", 'MIR4', 'Unturned', 'Terraria',
       'War Thunder', 'Hearts of Iron IV', 'Battlefield V', 'Valheim',
       'World of Tanks Blitz', 'Dayz', 'Total War: WARHAMMER II',
       "Don't Starve Together", 'Euro Truck Simulator 2', 'Rocket League',
       "Garry's Mod", 'Monster Hunter: World', 'Stardew Valley',
       'FINAL FANTASY XIV Online', 'Bless Unleashed ',
       'The Witcher 3: Wild Hunt', '雀魂麻将(MahjongSoul)', 'RimWorld',
       'Red Dead Redemption 2', '7 Days to Die', 'Black Desert',
       'Cities: Skylines', "Sid Meier's Civilization V",
       'Farming Simulator 19',

In [8]:
ID_list = monthly_players.URL.unique()

In [9]:
for i in range(0,len(ID_list)):
    ID_list[i] = ID_list[i].replace('https://steamcharts.com/app/', '')
ID_list

array(['730', '570', '578080', '1172470', '440', '271590', '1203220',
       '431960', '252490', '1085660', '230410', '1454400', '359550',
       '1263850', '218620', '346110', '381210', '289070', '1623660',
       '304930', '105600', '236390', '394360', '1238810', '892970',
       '444200', '221100', '594570', '322330', '227300', '252950', '4000',
       '582010', '413150', '39210', '1254120', '292030', '1329410',
       '294100', '1174180', '251570', '582660', '255710', '8930',
       '787860', '489830', '281990', '306130', '550', '1293830', '236850',
       '480', '813780', '1184370', '601510', '250900', '513710', '377160',
       '427520', '261550', '1281930', '1644960', '960090', '291550',
       '438100', '1097150', '107410', '1100600', '552990', '1313860',
       '374320', '1158310', '835570', '275850', '1259970', '1172620',
       '10', '740130', '238960', '629520', '1091500', '1129580',
       '1222680', '739630', '1225330', '1049590', '453480', '594650',
       '526870', '322

In [10]:
# Strip the URL data off the app ID
url_list = monthly_players["URL"]
for i in range(0,len(url_list)):
    url_list[i] = url_list[i].replace('https://steamcharts.com/app/', '')

C:\Users\David\AppData\Local\Temp/ipykernel_7864/1115660068.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  url_list[i] = url_list[i].replace('https://steamcharts.com/app/', '')


In [11]:
monthly_players = monthly_players.rename(columns={"URL": "APP"})
monthly_players

Month_Year  Avg_players       Gain Percent_Gain  Peak_Players  \
0     September 2021    512350.92     268.96       +0.05%        942519   
1        August 2021    512081.96    6014.60       +1.19%        802544   
2          July 2021    506067.36  -43279.72       -7.88%        763523   
3          June 2021    549347.08 -110541.81      -16.75%        929940   
4           May 2021    659888.89  -63457.63       -8.77%       1087197   
...              ...          ...        ...          ...           ...   
5266   December 2016       192.03     -21.90      -10.24%           405   
5267   November 2016       213.92    -134.68      -38.63%           537   
5268    October 2016       348.60    -201.75      -36.66%          1031   
5269  September 2016       550.36     543.26    +7654.89%          2836   
5270     August 2016         7.10        NaN          NaN            18   

         APP        Date                         Game_Name  
0        730  2021-09-01  Counter Strike: Global Offensive  
1        730  2021-08-01  Counter Strike: Global Offensive  
2        730  2021-07-01  Counter Strike: Global Offensive  
3        730  2021-06-01  Counter Strike: Global Offensive  
4        730  2021-05-01  Counter Strike: Global Offensive  
...      ...         ...                               ...  
5266  435150  2016-12-01          Divinity: Original Sin 2  
5267  435150  2016-11-01          Divinity: Original Sin 2  
5268  435150  2016-10-01          Divinity: Original Sin 2  
5269  435150  2016-09-01          Divinity: Original Sin 2  
5270  435150  2016-08-01          Divinity: Original Sin 2  

[5271 rows x 8 columns]

In [12]:
# split dataframe into relational dataframes
games = monthly_players[["APP", "Game_Name"]]
games = games.drop_duplicates().set_index("APP")
games

Game_Name
APP                                      
730      Counter Strike: Global Offensive
570                                Dota 2
578080                PUBG: Battlegrounds
1172470                      Apex Legends
440                       Team Fortress 2
...                                   ...
1238840                     Battlefield 1
239140                       Dying Light 
646570                     Slay the Spire
1222670                        The Sims 4
435150           Divinity: Original Sin 2

[98 rows x 1 columns]

In [13]:
reduced_players = monthly_players[["Percent_Gain", "APP", "Date"]]
reduced_players

Percent_Gain     APP        Date
0          +0.05%     730  2021-09-01
1          +1.19%     730  2021-08-01
2          -7.88%     730  2021-07-01
3         -16.75%     730  2021-06-01
4          -8.77%     730  2021-05-01
...           ...     ...         ...
5266      -10.24%  435150  2016-12-01
5267      -38.63%  435150  2016-11-01
5268      -36.66%  435150  2016-10-01
5269    +7654.89%  435150  2016-09-01
5270          NaN  435150  2016-08-01

[5271 rows x 3 columns]

## Remove bad rows

In [14]:
# Drop games with 1 or 2 rows; this is not enough data to extrapolate from, 
# as we will need to drop the first two months of a game's existence
game_id = reduced_players.APP.value_counts()
print(game_id)
clean_ID_list = []

for i in range(0,len(game_id)):
    if game_id[i] > 2:
        clean_ID_list.append(game_id.index[i])
    else:
        print(f"removing {games.loc[game_id.index[i]]}(app ID {game_id.index[i]}) due to having too few months of data")
        
len(clean_ID_list)

730        111
550        111
570        111
440        111
4000       111
          ... 
1203220      2
740130       1
1644960      1
1454400      1
1184370      1
Name: APP, Length: 98, dtype: int64
removing Game_Name    MIR4
Name: 1623660, dtype: object(app ID 1623660) due to having too few months of data
removing Game_Name    Bless Unleashed 
Name: 1254120, dtype: object(app ID 1254120) due to having too few months of data
removing Game_Name    NARAKA: BLADEPOINT
Name: 1203220, dtype: object(app ID 1203220) due to having too few months of data
removing Game_Name    Tales of Arise
Name: 740130, dtype: object(app ID 740130) due to having too few months of data
removing Game_Name    NBA 2k22
Name: 1644960, dtype: object(app ID 1644960) due to having too few months of data
removing Game_Name    Cookie Clicker
Name: 1454400, dtype: object(app ID 1454400) due to having too few months of data
removing Game_Name    Pathfinder: Wrath of Righteous
Name: 1184370, dtype: object(app ID 1184370)

91

In [15]:
# drop bottom two rows of each app, as the initial 2 months feature unusable data (NaN, followed by extreme outlier due to percentage calculation)
temp_df =pd.DataFrame(columns = reduced_players.columns)

for ID in clean_ID_list:
    df = reduced_players[reduced_players.APP == ID]
    df = df.reset_index(drop=True)
    print(df)
    df = df.drop(len(df)-1)
    df = df.drop(len(df)-1)
    print(df)
    temp_df = pd.concat([temp_df, df], ignore_index=True)
reduced_players = temp_df

    Percent_Gain  APP        Date
0         +0.05%  730  2021-09-01
1         +1.19%  730  2021-08-01
2         -7.88%  730  2021-07-01
3        -16.75%  730  2021-06-01
4         -8.77%  730  2021-05-01
..           ...  ...         ...
106      +31.61%  730  2012-11-01
107      -32.88%  730  2012-10-01
108       +3.39%  730  2012-09-01
109    +1559.43%  730  2012-08-01
110          NaN  730  2012-07-01

[111 rows x 3 columns]
    Percent_Gain  APP        Date
0         +0.05%  730  2021-09-01
1         +1.19%  730  2021-08-01
2         -7.88%  730  2021-07-01
3        -16.75%  730  2021-06-01
4         -8.77%  730  2021-05-01
..           ...  ...         ...
104      +14.81%  730  2013-01-01
105       -0.39%  730  2012-12-01
106      +31.61%  730  2012-11-01
107      -32.88%  730  2012-10-01
108       +3.39%  730  2012-09-01

[109 rows x 3 columns]
    Percent_Gain  APP        Date
0        -14.13%  550  2021-09-01
1         -0.46%  550  2021-08-01
2        +20.23%  550  2021-07-01


[85 rows x 3 columns]
   Percent_Gain     APP        Date
0       -16.05%  250900  2021-09-01
1       -12.95%  250900  2021-08-01
2       -21.30%  250900  2021-07-01
3       -33.62%  250900  2021-06-01
4       -25.42%  250900  2021-05-01
..          ...     ...         ...
78      -13.64%  250900  2015-03-01
79      -26.18%  250900  2015-02-01
80      -17.33%  250900  2015-01-01
81      -30.90%  250900  2014-12-01
82          NaN  250900  2014-11-01

[83 rows x 3 columns]
   Percent_Gain     APP        Date
0       -16.05%  250900  2021-09-01
1       -12.95%  250900  2021-08-01
2       -21.30%  250900  2021-07-01
3       -33.62%  250900  2021-06-01
4       -25.42%  250900  2021-05-01
..          ...     ...         ...
76       -2.28%  250900  2015-05-01
77      -17.31%  250900  2015-04-01
78      -13.64%  250900  2015-03-01
79      -26.18%  250900  2015-02-01
80      -17.33%  250900  2015-01-01

[81 rows x 3 columns]
   Percent_Gain     APP        Date
0        -3.36%  322170  2021-09

59          NaN  453480  2016-10-01
   Percent_Gain     APP        Date
0        +7.08%  453480  2021-09-01
1       -38.28%  453480  2021-08-01
2        +8.94%  453480  2021-07-01
3       +35.62%  453480  2021-06-01
4       -14.43%  453480  2021-05-01
5       +40.69%  453480  2021-04-01
6        -9.62%  453480  2021-03-01
7       -31.01%  453480  2021-02-01
8       +18.75%  453480  2021-01-01
9       +17.96%  453480  2020-12-01
10      -16.50%  453480  2020-11-01
11      +12.24%  453480  2020-10-01
12       -9.81%  453480  2020-09-01
13      -28.91%  453480  2020-08-01
14      +17.49%  453480  2020-07-01
15      +11.22%  453480  2020-06-01
16      -19.91%  453480  2020-05-01
17      +46.25%  453480  2020-04-01
18       +1.36%  453480  2020-03-01
19      -20.29%  453480  2020-02-01
20      +16.61%  453480  2020-01-01
21      +30.19%  453480  2019-12-01
22      -22.65%  453480  2019-11-01
23      -14.94%  453480  2019-10-01
24      +15.92%  453480  2019-09-01
25      -32.42%  453480  201

15          NaN  1172620  2020-06-01
   Percent_Gain      APP        Date
0        -7.39%  1172620  2021-09-01
1       -48.84%  1172620  2021-08-01
2       -11.06%  1172620  2021-07-01
3       +81.27%  1172620  2021-06-01
4       -27.69%  1172620  2021-05-01
5       +74.02%  1172620  2021-04-01
6       -28.95%  1172620  2021-03-01
7       -30.09%  1172620  2021-02-01
8       +62.68%  1172620  2021-01-01
9       +23.92%  1172620  2020-12-01
10      +70.63%  1172620  2020-11-01
11      -35.27%  1172620  2020-10-01
12      -21.66%  1172620  2020-09-01
13      -43.33%  1172620  2020-08-01
   Percent_Gain      APP        Date
0       +15.63%  1222680  2021-09-01
1       -21.40%  1222680  2021-08-01
2       -32.67%  1222680  2021-07-01
3      +181.38%  1222680  2021-06-01
4       +52.86%  1222680  2021-05-01
5       -39.06%  1222680  2021-04-01
6       -38.02%  1222680  2021-03-01
7       -26.39%  1222680  2021-02-01
8       +58.79%  1222680  2021-01-01
9       +11.36%  1222680  2020-12-01
1

In [16]:
# reprint the cleaned up reduced_players
reduced_players

# nb: perhaps the above cleaning process could have been done in one step by simply catching and handling errors?

Percent_Gain      APP        Date
0          +0.05%      730  2021-09-01
1          +1.19%      730  2021-08-01
2          -7.88%      730  2021-07-01
3         -16.75%      730  2021-06-01
4          -8.77%      730  2021-05-01
...           ...      ...         ...
5074       -3.43%  1293830  2021-09-01
5075      -47.45%  1293830  2021-08-01
5076      +68.25%  1293830  2021-07-01
5077      +17.89%  1293830  2021-06-01
5078      -36.69%  1293830  2021-05-01

[5079 rows x 3 columns]

## Fix Datatypes


In [17]:
# strip off the percent symbol and convert it into a float
percentages = reduced_players["Percent_Gain"]
for i in range(0, len(reduced_players)):
    try:
        reduced_players.loc[i, "Percent_Gain"] = float(reduced_players.loc[i, "Percent_Gain"].replace("%",""))
    except AttributeError as e:
        app = reduced_players.loc[i, "APP"]
        print(f"error on app id {app}")
        print(e)

print(reduced_players.dtypes)

Percent_Gain    object
APP             object
Date            object
dtype: object


In [18]:
reduced_players

Percent_Gain      APP        Date
0            0.05      730  2021-09-01
1            1.19      730  2021-08-01
2           -7.88      730  2021-07-01
3          -16.75      730  2021-06-01
4           -8.77      730  2021-05-01
...           ...      ...         ...
5074        -3.43  1293830  2021-09-01
5075       -47.45  1293830  2021-08-01
5076        68.25  1293830  2021-07-01
5077        17.89  1293830  2021-06-01
5078       -36.69  1293830  2021-05-01

[5079 rows x 3 columns]

In [19]:
# check all percent values are now floats:
for i in range (0,len(reduced_players)):
    if(type(reduced_players.loc[i][0]) != float):
        print("non-float found")

In [20]:
# we can keep APP as a str as we don't want to be using it to make any predictions.

In [21]:
# convert 'Date' to true date-time format
for i in range (0,len(reduced_players)):
    reduced_players.loc[i][2] = pd.to_datetime(reduced_players.loc[i][2])

In [22]:
reduced_players

Percent_Gain      APP                 Date
0            0.05      730  2021-09-01 00:00:00
1            1.19      730  2021-08-01 00:00:00
2           -7.88      730  2021-07-01 00:00:00
3          -16.75      730  2021-06-01 00:00:00
4           -8.77      730  2021-05-01 00:00:00
...           ...      ...                  ...
5074        -3.43  1293830  2021-09-01 00:00:00
5075       -47.45  1293830  2021-08-01 00:00:00
5076        68.25  1293830  2021-07-01 00:00:00
5077        17.89  1293830  2021-06-01 00:00:00
5078       -36.69  1293830  2021-05-01 00:00:00

[5079 rows x 3 columns]

## Read in news data

In [24]:
# this time, I am going to wrap the whole thing in a for loop, to do a separate call for each game.

Monthly_data_cleaned_joined = pd.DataFrame()

for app_no in clean_ID_list:
    print(f"new app:{app_no}")
    if (app_no == '480'):
        # UPDATE! quick hack:
        # skip app 480 - it is a secret test game and has no news feed.
        continue
    current_df = reduced_players[reduced_players.APP == app_no]
    current_df = current_df.reset_index(drop=True)
#     print(current_df)
    
    # build dataframe from json data and reduce to relevant columns

    # call for just one news item
    count = 1
    max_length = 1024
    api_url = f"http://api.steampowered.com/ISteamNews/GetNewsForApp/v0002/?appid={app_no}&count={count}&maxlength={max_length}&format=json"
    news_response = requests.get(api_url).json()

    # find out how many news items for this app, and then recall news items.
    count = news_response['appnews']['count']
    api_url = f"http://api.steampowered.com/ISteamNews/GetNewsForApp/v0002/?appid={app_no}&count={count}&maxlength={max_length}&format=json"
    news_response = requests.get(api_url).json()
    
    # new DF to store response
    newsDF = pd.DataFrame(columns=['Date', 'Label'])

    # get the date and label of each news item; we can ignore the actual text of the news item on this first approach.
    for item in (news_response['appnews']['newsitems']):
        # convert unix date to yyy-mm-dd
#         print(f"{pd.to_datetime(item['date'], unit='s')} {item['feedlabel']}")
        row = pd.DataFrame(data={'Date': [pd.to_datetime(item['date'], unit='s')], 'Label': [item['feedlabel']]})
        newsDF = pd.concat([newsDF, row])
    
    #reset the index and print out the important values from the dataframe
    newsDF = newsDF.reset_index(drop=True)
#     print(newsDF)
    print(newsDF.nunique())
    print(newsDF.Label.value_counts())
    
    # add new label columns (if any) to the dataframe
    labels_index = newsDF.Label.value_counts().index
    for column in labels_index:
        current_df[column] = 0
    current_df
#     print(current_df)
    
    # bin news feed into monthly groups
    for i in range(0, len(current_df)):
        # set start and end dates for the current month
        start_date = current_df.loc[i, 'Date']
        end_date = start_date + relativedelta(months = 1)
#         print(f"{start_date}, {end_date}")
        #identify rows from the newsDF that fall into that category
        for j in range(0, len(newsDF)):
            if ((newsDF.loc[j, 'Date'] >= start_date) & (newsDF.loc[j, 'Date'] < end_date)):
                # find the relevant label and increment that column on the overall grid
                feed_type = newsDF.loc[j, 'Label']
#                 print(feed_type)
                current_df.loc[i, feed_type] += 1
    Monthly_data_cleaned_joined = pd.concat([Monthly_data_cleaned_joined, current_df], ignore_index=True)

new app:730
Date     1265
Label      12
dtype: int64
PC Gamer                   434
Rock, Paper, Shotgun       259
The Loadout                150
Community Announcements    138
SteamDB                    127
PCGamesN                    79
Product Update              72
Eurogamer                   23
GamingOnLinux               15
Announcement                 9
Kotaku                       6
Product Release              3
Name: Label, dtype: int64
new app:550
Date     812
Label     13
dtype: int64
L4D Blog                   272
Community Announcements    268
Product Update             165
PC Gamer                    82
Kotaku                      65
Rock, Paper, Shotgun        50
Shacknews                   44
Eurogamer                   18
Announcement                14
GamingOnLinux                5
PCGamesN                     3
Product Release              2
VG247                        1
Name: Label, dtype: int64
new app:570
Date     2700
Label      16
dtype: int64
Product Update  

new app:322330
Date     386
Label      5
dtype: int64
Community Announcements    541
GamingOnLinux               10
Announcement                 4
Rock, Paper, Shotgun         3
Product Release              1
Name: Label, dtype: int64
new app:239140
Date     291
Label      8
dtype: int64
Community Announcements    262
Rock, Paper, Shotgun        48
Announcement                11
GamingOnLinux                9
VG247                        6
Eurogamer                    5
Product Release              4
PCGamesN                     2
Name: Label, dtype: int64
new app:255710
Date     552
Label      8
dtype: int64
Community Announcements    392
PC Gamer                    96
Rock, Paper, Shotgun        90
PCGamesN                    21
Eurogamer                   11
Announcement                 8
Product Release              5
GamingOnLinux                4
Name: Label, dtype: int64
new app:271590
Date     1113
Label      10
dtype: int64
PC Gamer                   407
Rock, Paper, Shotgun  

new app:594570
Date     531
Label      9
dtype: int64
Community Announcements    419
Rock, Paper, Shotgun        76
PC Gamer                    65
PCGamesN                    29
Strategy Gamer              26
Eurogamer                   18
GamingOnLinux                6
Announcement                 2
Product Release              1
Name: Label, dtype: int64
new app:582660
Date     743
Label      6
dtype: int64
Community Announcements    868
Announcement                 4
PCGamesN                     3
Rock, Paper, Shotgun         3
Eurogamer                    1
Product Release              1
Name: Label, dtype: int64
new app:646570
Date     193
Label      8
dtype: int64
Community Announcements    120
Rock, Paper, Shotgun        52
PC Gamer                    29
Eurogamer                   11
GamingOnLinux                3
PCGamesN                     2
VG247                        1
Announcement                 1
Name: Label, dtype: int64
new app:513710
Date     355
Label      5
dtype:

new app:1263850
Date     79
Label     6
dtype: int64
Community Announcements    52
PCGamesN                   10
Eurogamer                   8
SteamDB                     6
Rock, Paper, Shotgun        2
VG247                       1
Name: Label, dtype: int64
new app:1091500
Date     921
Label     10
dtype: int64
Gamemag.ru                 362
Rock, Paper, Shotgun       241
PC Gamer                   157
Eurogamer                   66
Community Announcements     55
SteamDB                     31
VG247                       23
Shacknews                   11
pressakey.com                6
Announcement                 1
Name: Label, dtype: int64
new app:1366540
Date     103
Label      5
dtype: int64
Community Announcements    89
PCGamesN                    8
SteamDB                     4
Rock, Paper, Shotgun        3
Strategy Gamer              1
Name: Label, dtype: int64
new app:892970
Date     173
Label      7
dtype: int64
Community Announcements    62
GamingOnLinux              38
Eurog

In [ ]:
Monthly_data_cleaned_joined = Monthly_data_cleaned_joined.fillna(0)

In [ ]:
# add binary target: positive or negative change
# 1 for positive, 0 for negative
Monthly_data_cleaned_joined["Binary_outcome"] = 0
Monthly_data_cleaned_joined

In [ ]:
for i in range (0, len(Monthly_data_cleaned_joined)):
    if (Monthly_data_cleaned_joined.loc[i, 'Percent_Gain'] > 0):
        Monthly_data_cleaned_joined.loc[i, 'Binary_outcome'] = 1
    elif (Monthly_data_cleaned_joined.loc[i, 'Percent_Gain'] < 0):
        Monthly_data_cleaned_joined.loc[i, 'Binary_outcome'] = 0
    else:
        print("this should never be printed; true 0 found")

Monthly_data_cleaned_joined

# Machine Learning

In [ ]:
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans
from sklearn.manifold import TSNE

path = "Resources/myopia.csv"
from pathlib import Path


from sklearn.datasets import make_regression
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier

from sklearn.metrics import classification_report, confusion_matrix

import tensorflow as tf

### supervised

In [ ]:
# drop non-numerical data
Monthly_data_cleaned_joined = Monthly_data_cleaned_joined.drop(columns=['APP', 'Date', 'Percent_Gain'])

In [ ]:
# make the choice of randomiser explicit:
sample_hold = 42

In [ ]:
Monthly_data_cleaned_joined.shape

In [ ]:
# Split the data into X_train, X_test, y_train, y_test
y = Monthly_data_cleaned_joined['Binary_outcome']

X = Monthly_data_cleaned_joined.drop(columns=['Binary_outcome'])


print("Shape: ", X.shape, y.shape)



In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=sample_hold)



In [1]:
scaler = StandardScaler().fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)
Percent_Gain



NameError: name 'StandardScaler' is not defined

In [ ]:
# Train a Logistic Regression model print the model score
classifier = LogisticRegression()
classifier.fit(X_train_scaled, y_train)



In [ ]:
print(f"Training Data Score: {classifier.score(X_train_scaled, y_train)}")
print(f"Testing Data Score: {classifier.score(X_test_scaled, y_test)}")



In [ ]:
target_names = ["0", "1"]

In [ ]:
clf = RandomForestClassifier(random_state=sample_hold).fit(X_train, y_train)
y_pred = clf.predict(X_test)
print(classification_report(y_test, y_pred, target_names=target_names))
print(f'Training Score: {clf.score(X_train, y_train)}')
print(f'Testing Score: {clf.score(X_test, y_test)}')



### unsupervised

In [ ]:
Monthly_data_cleaned_joined.shape

In [ ]:
Monthly_data_cleaned_joined.describe()

In [ ]:
pd.plotting.scatter_matrix(Monthly_data_cleaned_joined, c=Monthly_data_cleaned_joined["Binary_outcome"], figsize=(16,16))
plt.savefig("1.png", format="png")



In [ ]:
Monthly_data_cleaned_joined_unsuper = Monthly_data_cleaned_joined.drop(columns=["Binary_outcome"])
Monthly_data_cleaned_joined_unsuper



In [ ]:
scaler = StandardScaler()


X = pd.DataFrame(scaler.fit_transform(X))
X



In [ ]:
# Initialize PCA model for 90% explained variance
pca = PCA(n_components=0.9)

# Get two principal components for the data.
pcaX = pd.DataFrame(pca.fit_transform(X))
pcaX



In [ ]:
tsne = TSNE(learning_rate=250)
tsne_features = tsne.fit_transform(pcaX)
TSNE_DF = pd.DataFrame(tsne_features)
TSNE_DF = TSNE_DF.rename(columns={0: "X", 1: "Y"})

TSNE_DF


In [ ]:
plt.scatter(TSNE_DF.X, TSNE_DF.Y)



In [ ]:
inertia = []
k = list(range(1, 11))

# Calculate the inertia for the range of k values
for i in k:
    km = KMeans(n_clusters=i, random_state=0)
    km.fit(TSNE_DF)
    inertia.append(km.inertia_)

# Create the Elbow Curve using hvPlot
elbow_data = {"k": k, "inertia": inertia}
df_elbow = pd.DataFrame(elbow_data)
df_elbow

In [ ]:
# Plot the elbow curve to find the best candidate(s) for k
plt.plot(df_elbow['k'], df_elbow['inertia'])
plt.xticks(range(1,11))
plt.xlabel('Number of clusters')
plt.ylabel('Inertia')
plt.title('Elbow curve for Myopia')
plt.show()



In [ ]:
def get_clusters(k, data):
    # Initialize the K-Means model
    model = KMeans(n_clusters=k, random_state=0)

    # Train the model
    model.fit(data)

    # Predict clusters
    predictions = model.predict(data)

    # Create return DataFrame with predicted clusters
    data["class"] = model.labels_

    return data

In [ ]:
clusters = get_clusters(4, TSNE_DF)

In [ ]:
clusters

In [ ]:
plt.scatter(x=clusters.X, y=clusters.Y, c=clusters["class"])



In [ ]:
plt.scatter(TSNE_DF.X, TSNE_DF.Y, c=Monthly_data_cleaned_joined.Binary_outcome)



### artificial neural network

In [ ]:
input_dimensions=X.shape[1]
input_dimensions

In [ ]:
X

In [ ]:
# Create a method that creates a new Sequential model with hyperparameter options
def create_model(hp):
    nn_model = tf.keras.models.Sequential()

    # Allow kerastuner to decide which activation function to use in hidden layers
    activation = hp.Choice('activation',['relu','tanh'])
    
    # Allow kerastuner to decide number of neurons in first layer
    nn_model.add(tf.keras.layers.Dense(units=hp.Int('first_units',
        min_value=1,
        max_value=30,
        step=5), activation=activation, input_dim=input_dimensions))

    # Allow kerastuner to decide number of hidden layers and neurons in hidden layers
    for i in range(hp.Int('num_layers', 1, 5)):
        nn_model.add(tf.keras.layers.Dense(units=hp.Int('units_' + str(i),
            min_value=1,
            max_value=100,
            step=5),
            activation=activation))
    
    nn_model.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

    # Compile the model
    nn_model.compile(loss="binary_crossentropy", optimizer='adam', metrics=["accuracy"])
    
    return nn_model

In [ ]:
# Import the kerastuner library
import keras_tuner as kt

tuner = kt.Hyperband(
    create_model,
    objective="val_accuracy",
    overwrite=True,
    max_epochs=20,
    hyperband_iterations=2,)

In [ ]:
# Run the kerastuner search for best hyperparameters
tuner.search(X_train_scaled,y_train,epochs=20,validation_data=(X_test_scaled,y_test))

### Search has not yielded anything higher than 55% accuracy. 
I will now experiment with dropping more columns from the data.